### PR TITLE
iotexAdd unit tests for action/protocol/poll/protocol_test.go #638

### PR DESCRIPTION
**Pull request by lzxm160**
work on #638

### CODE REVIEW

File: action/protocol/poll/protocol_test.go [hunk 1]
```diff
@@ -24,12 +31,18 @@ import (
 	"github.com/iotexproject/iotex-core/state/factory"
 )
 
-func TestInitialize(t *testing.T) {
+func initConstruct(t *testing.T)(Protocol,context.Context,factory.WorkingSet,*types.ElectionResult){
```
Comment (CoderZhi) [inline action/protocol/poll/protocol_test.go]: separate return values with ", "

File: action/protocol/poll/protocol_test.go [hunk 2]
```diff
@@ -52,6 +66,11 @@ func TestInitialize(t *testing.T) {
 		cfg.Genesis.NumDelegates,
 	)
 	require.NoError(err)
+	return p,ctx,ws,r
+}
+func TestInitialize(t *testing.T) {
```
Comment (CoderZhi) [inline action/protocol/poll/protocol_test.go]: empty line above

File: action/protocol/poll/protocol_test.go [hunk 3]
```diff
@@ -52,6 +66,11 @@ func TestInitialize(t *testing.T) {
 		cfg.Genesis.NumDelegates,
 	)
 	require.NoError(err)
+	return p,ctx,ws,r
```
Comment (CoderZhi) [inline action/protocol/poll/protocol_test.go]: return p, ctx, ws, r

File: action/protocol/poll/protocol_test.go [hunk 4]
```diff
@@ -52,6 +66,11 @@ func TestInitialize(t *testing.T) {
...
+}
+func TestInitialize(t *testing.T) {
+	require := require.New(t)
+	p,ctx,ws,r:=initConstruct(t)
```
Comment (CoderZhi) [inline action/protocol/poll/protocol_test.go]: spaces

File: action/protocol/poll/protocol_test.go [hunk 5]
```diff
@@ -69,7 +88,185 @@ func TestInitialize(t *testing.T) {
...
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	sm := mock_chainmanager.NewMockStateManager(ctrl)
+	cb := db.NewCachedBatch()
+	sm.EXPECT().GetCachedBatch().Return(cb).AnyTimes()
+	sm.EXPECT().State(gomock.Any(), gomock.Any()).Return(state.ErrStateNotExist).AnyTimes()
+	sm.EXPECT().PutState(gomock.Any(), gomock.Any()).Return(nil).AnyTimes()
+	sm.EXPECT().Snapshot().Return(1).AnyTimes()
+	p,ctx,ws,_:=initConstruct(t)
```
Comment (CoderZhi) [inline action/protocol/poll/protocol_test.go]: spaces

File: action/protocol/poll/protocol_test.go [hunk 6]
```diff
@@ -69,7 +88,185 @@ func TestInitialize(t *testing.T) {
...
+	require.NoError(err)
+	bd := &action.EnvelopeBuilder{}
+	elp := bd.SetGasLimit(uint64(100000)).
+		SetGasPrice(big.NewInt(10)).
+		SetAction(tsf).Build()
+	selp, err := action.Sign(elp, senderKey.PriKey)
+	require.NoError(err)
+	require.NotNil(selp)
+	// Case 1: wrong action type
+	receipt,err:=p.Handle(ctx,selp.Action(),nil)
```
Comment (CoderZhi) [inline action/protocol/poll/protocol_test.go]: spaces

File: action/protocol/poll/protocol_test.go [hunk 7]
```diff
@@ -69,7 +88,185 @@ func TestInitialize(t *testing.T) {
...
+	require.Nil(err)
```
Comment (CoderZhi) [inline action/protocol/poll/protocol_test.go]: require.NoError

File: action/protocol/poll/protocol_test.go [hunk 8]
```diff
@@ -69,7 +88,185 @@ func TestInitialize(t *testing.T) {
...
+	var sc state.CandidateList
+	require.NoError(ws.State(candidatesutil.ConstructKey(1), &sc))
+	act := action.NewPutPollResult(1, 123456, sc)
+	elp = bd.SetGasLimit(uint64(100000)).
+		SetGasPrice(big.NewInt(10)).
+		SetAction(act).Build()
+	selp, err = action.Sign(elp, senderKey.PriKey)
+	require.NoError(err)
+	require.NotNil(selp)
+	receipt,err=p.Handle(ctx,selp.Action(),sm)
```
Comment (CoderZhi) [inline action/protocol/poll/protocol_test.go]: spaces

File: action/protocol/poll/protocol_test.go [hunk 9]
```diff
@@ -69,7 +88,185 @@ func TestInitialize(t *testing.T) {
...
+	require.Error(err)
+	require.Nil(receipt)
+	// Case 3: all right
+	p3,ctx3,ws3,_:=initConstruct(t)
```
Comment (CoderZhi) [inline action/protocol/poll/protocol_test.go]: spaces

File: action/protocol/poll/protocol_test.go [hunk 10]
```diff
@@ -69,7 +88,185 @@ func TestInitialize(t *testing.T) {
...
+	var sc3 state.CandidateList
+	require.NoError(ws3.State(candidatesutil.ConstructKey(1), &sc3))
+	act3 := action.NewPutPollResult(1, 1, sc3)
+	elp = bd.SetGasLimit(uint64(100000)).
+		SetGasPrice(big.NewInt(10)).
+		SetAction(act3).Build()
+	selp3, err := action.Sign(elp, senderKey.PriKey)
+	require.NoError(err)
+	require.NotNil(selp3)
+	receipt,err=p.Handle(ctx3,selp3.Action(),sm)
```
Comment (CoderZhi) [inline action/protocol/poll/protocol_test.go]: spaces

File: action/protocol/poll/protocol_test.go [hunk 11]
```diff
@@ -69,7 +88,185 @@ func TestInitialize(t *testing.T) {
...
+	require.NoError(err)
+	require.NotNil(receipt)
 }
 
 func TestProtocol_Validate(t *testing.T) {
+	require := require.New(t)
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	p,ctx,ws,_:=initConstruct(t)
```
Comment (CoderZhi) [inline action/protocol/poll/protocol_test.go]: spaces

File: action/protocol/poll/protocol_test.go [hunk 12]
```diff
@@ -69,7 +88,185 @@ func TestInitialize(t *testing.T) {
...
+	require.NoError(err)
+	bd := &action.EnvelopeBuilder{}
+	elp := bd.SetGasLimit(uint64(100000)).
+		SetGasPrice(big.NewInt(10)).
+		SetAction(tsf).Build()
+	selp, err := action.Sign(elp, senderKey.PriKey)
+	require.NoError(err)
+	require.NotNil(selp)
+	// Case 1: wrong action type
+	err=p.Validate(ctx,selp.Action())
```
Comment (CoderZhi) [inline action/protocol/poll/protocol_test.go]: require.NoError(p.Validate(ctx, selp.Action()))

File: action/protocol/poll/protocol_test.go [hunk 13]
```diff
@@ -69,7 +88,185 @@ func TestInitialize(t *testing.T) {
...
+	require.Nil(err)
+	// Case 2: Only producer could create this protocol
+	p2,ctx2,ws2,_:=initConstruct(t)
```
Comment (CoderZhi) [inline action/protocol/poll/protocol_test.go]: spaces

File: action/protocol/poll/protocol_test.go [hunk 14]
```diff
@@ -69,7 +88,185 @@ func TestInitialize(t *testing.T) {
...
+	require.NoError(err)
+	ctx2 = protocol.WithValidateActionsCtx(
+		context.Background(),
+		protocol.ValidateActionsCtx{
+			BlockHeight:  1,
+			ProducerAddr: recipientAddr.String(),
+			Caller:       caller,
+		},
+	)
+	err=p.Validate(ctx2,selp2.Action())
```
Comment (CoderZhi) [inline action/protocol/poll/protocol_test.go]: spaces

File: action/protocol/poll/protocol_test.go [hunk 15]
```diff
@@ -69,7 +88,185 @@ func TestInitialize(t *testing.T) {
...
+	require.True(strings.Contains(err.Error(), "Only producer could create this protocol"))
+	// Case 3: duplicate candidate
+	p3,ctx3,ws3,_:=initConstruct(t)
+	require.NoError(p3.Initialize(ctx3, ws3))
+	var sc3 state.CandidateList
+	require.NoError(ws3.State(candidatesutil.ConstructKey(1), &sc3))
+	sc3=append(sc3,&state.Candidate{"1",big.NewInt(10),"2"})
```
Comment (CoderZhi) [inline action/protocol/poll/protocol_test.go]: spaces

File: action/protocol/poll/protocol_test.go [hunk 16]
```diff
@@ -69,7 +88,185 @@ func TestInitialize(t *testing.T) {
...
+	sc3=append(sc3,&state.Candidate{"1",big.NewInt(10),"2"})
```
Comment (CoderZhi) [inline action/protocol/poll/protocol_test.go]: spaces

File: action/protocol/poll/protocol_test.go [hunk 17]
```diff
@@ -69,7 +88,185 @@ func TestInitialize(t *testing.T) {
...
+		context.Background(),
+		protocol.ValidateActionsCtx{
+			BlockHeight:  1,
+			ProducerAddr: testaddress.Addrinfo["producer"].String(),
+			Caller:       caller,
+		},
+	)
+	err=p4.Validate(ctx4,selp4.Action())
+	require.True(strings.Contains(err.Error(), "the proposed delegate list length"))
+
```
Comment (CoderZhi) [inline action/protocol/poll/protocol_test.go]: add a case that the candidate's weight or something else are not identical

Comment (lzxm160) [reply action/protocol/poll/protocol_test.go]: ok

**Pull request closed without merging**